bayem/vba.py

```diff
@@ -0,0 +1,553 @@
+import logging
+from collections import defaultdict
+from dataclasses import dataclass
+from time import perf_counter
+from tabulate import tabulate
+from typing import Union, Dict, Tuple
+
+import numpy as np
+import scipy.special as special
+
+from .distributions import MVN, Gamma
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class VBOptions:
+    tolerance: float = 0.1
+    maxiter: int = 50
+    maxtrials: int = 10
+    update_noise: Union[Dict, bool] = True
+    index_ARD: Tuple[int] = ()
+
+    # epsilon for central differences jacobian,  approx sqrt(machine precision):
+    cdf_eps: float = np.finfo(np.float).eps ** 0.5
+
+    # If true, includes the full precision for each iteration in the VBResult.
+    # Set that to False for _big_ problems to save memory.
+    store_full_precision: bool = True
+
+
+class CDF_Jacobian:
+    def __init__(self, f, transformation, cdf_eps):
+        """
+        Provides a numerical jacobian df/dtheta for the user-provided callable
+        `f` based on central differences.
+
+        f:
+            user provided callable, see `bayem.vba`
+
+        transformation:
+            callable that turns the output of `f` into the `dict` format
+            of the `VBAProblem` class. See the "additional notes" section
+            in `bayem.vba`
+
+        cdf_eps:
+            defines the step length of parameter `x` as 
+                h = max(cdf_eps, abs(x) * cdf_eps)
+        """
+        self._f = f
+        self._T = transformation
+        self.cdf_eps = cdf_eps
+
+    def __call__(self, _x):
+        x = np.copy(_x)
+
+        for iParam in range(len(x)):
+            eps = self.cdf_eps
+            dx = max(eps, abs(x[iParam]) * eps)
+
+            x[iParam] -= dx
+            fs0 = self._T(self._f(x))
+            x[iParam] += 2 * dx
+            fs1 = self._T(self._f(x))
+            x[iParam] = _x[iParam]
+
+            if iParam == 0:
+                # allocate jac
+                jac = {}
+                for key, f0 in fs0.items():
+                    jac[key] = np.empty([len(f0), len(x)])
+
+            for n in fs0:
+                jac[n][:, iParam] = (fs1[n] - fs0[n]) / (2 * dx)
+
+        return jac
+
+
+class VBAProblem:
+    def __init__(self, f, noise0, jac, options):
+        self.f = f
+        self.noise0 = noise0
+        self.jac = jac
+        self.jac_in_f = not callable(jac) and jac
+        self.options = options
+
+        self._Tk = None  # transformation of k = f(x)
```

Comment:
What transformation is meant here (and also for the next line of the code: self._TJ)? Is it by default no (identity) transformation?

Reply:
I renamed the class to `DictModelError` and added a docstring with more information. See, if that is clearer now.

Reply:
I see now. I had thought , it would be related to transformation of the qunatities (numbers themselves) , but it is only about the structure for representing them. Alright, is now clear.